### PR TITLE
 [HW-3] Solve 4 tasks 

### PR DESCRIPTION
**Pull request by Captain-Vorobey**
Results of default input data:
image

### CODE REVIEW

File: HW03/vorobyov_egor/index.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+require_relative './lib/logs_parser'
+
+p '-------------------------------------------------------------------------------------------------------------------------------------------'
+
+puts LogsParser.task_1('10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0498',
+                       '10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] "POST /grid/2/event HTTP/1.1" 200 - 0.2277',
+                       '2019-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>',
+                       '10.6.246.101 - - [23/Apr/2018:20:29:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0290')
```
Comment (B0RlS) [inline HW03/vorobyov_egor/index.rb:8]:
```suggestion
logs = <<~ HEREDOC
  10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0498
  10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] "POST /grid/2/event HTTP/1.1" 200 - 0.2277
  2018-04-23 20:30:42: SSL ErRor, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>
  10.6.246.101 - - [23/Apr/2018:20:29:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0290
HEREDOC


puts LogsParser.task_1(logs)
```

File: HW03/vorobyov_egor/lib/predicates.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+module Predicates
+  def self.has_error?(el)
+    el if el.include? 'error'
```
Comment (B0RlS) [inline HW03/vorobyov_egor/lib/predicates.rb:3]: it must return true or false

File: HW03/vorobyov_egor/lib/parsing_functions.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
+require 'time'
+
+module ParsingFunctions
+  def self.formatted(el)
+    date = el.match(%r{\d{2}/\w{3}/\d{4}:\d{2}:\d{2}:\d{2} \+\d{4}}).to_s
+    ip = el.match(/\d{2}\.\d\.\d{3}\.\d{3}/).to_s
+    address = el.match(%r{\s/\w{1,}/\d/\w{1,}}).to_s
```
Comment (B0RlS) [inline HW03/vorobyov_egor/lib/parsing_functions.rb:7]: Could you please move regex to constants?

File: HW03/vorobyov_egor/lib/logs_parser.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    end
+    arr
+  end
+
+  def self.task_3(logs)
+    0 if logs.length < 2
+
+    res = []
+    new_el = ''
+    logs.each_with_index do |_el, i|
```
Comment (B0RlS) [inline HW03/vorobyov_egor/lib/logs_parser.rb:28]: where are you using _el?

File: HW03/vorobyov_egor/lib/parsing_functions.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+    date + ' FROM: ' + ip + ' TO:' + address.upcase
+  end
+
+  def self.calling_core_logs(logs)
+    new_arr = []
+    logs.each do |el|
+      if el.include? 'Calling core with action:'
+        date = el.split(' ubuntu')[0]
```
Comment (B0RlS) [inline HW03/vorobyov_egor/lib/parsing_functions.rb:16]: But what will happen if there is no ubuntu?

File: HW03/vorobyov_egor/lib/predicates.rb (hunk 2 above)
Comment (vladhilko) [reply HW03/vorobyov_egor/lib/predicates.rb:3]: @Captain-Vorobey still actual

File: HW03/vorobyov_egor/lib/predicates.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+module Predicates
+  def self.with_error(el)
+    el if el.include? 'error'
```
Comment (vladhilko) [inline HW03/vorobyov_egor/lib/predicates.rb:3]: We usually don't create module just for one line at all 😅 
```suggestion
  def self.with_error?(el)
    el.include? 'error'
```

File: HW03/vorobyov_egor/lib/logs_parser.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    arr
+  end
+
+  def self.task_3(logs)
+    0 if logs.length < 2
+
+    res = []
+    logs.each_with_index do |_el, i|
+      _el = ParsingFunctions.dates_difference(logs[i], logs[i + 1]) if logs[i + 1] != nil
+      res.push(_el)
```
Comment (vladhilko) [inline HW03/vorobyov_egor/lib/logs_parser.rb:29]: There should be something like this 
```suggestion
    logs.each_with_index do |element, index|
      date_diffirence = ParsingFunctions.dates_difference(element, logs[index + 1]) unless logs[index+ 1].empty?
      res.push(date_diffirence)
```

Comment (Captain-Vorobey) [reply HW03/vorobyov_egor/lib/logs_parser.rb]: I can't use `emty?` method for `nil:NilClass`

**Pull request merged**